drivers/FlashIAP.cpp

```diff
@@ -107,50 +109,58 @@ int FlashIAP::program(const void *buffer, uint32_t addr, uint32_t size)
     uint32_t page_size = get_page_size();
     uint32_t flash_size = flash_get_size(&_flash);
     uint32_t flash_start_addr = flash_get_start_address(&_flash);
-    uint32_t chunk, prog_size;
-    const uint8_t *buf = (uint8_t *) buffer;
-    const uint8_t *prog_buf;
+    int ret;
 
     // addr should be aligned to page size
     if (!is_aligned(addr, page_size) || (!buffer) ||
             ((addr + size) > (flash_start_addr + flash_size))) {
         return -1;
     }
 
-    int ret = 0;
     _mutex->lock();
-    while (size) {
-        uint32_t current_sector_size = flash_get_sector_size(&_flash, addr);
-        bool unaligned_src = (((size_t) buf / sizeof(uint32_t) * sizeof(uint32_t)) != (size_t) buf);
-        chunk = std::min(current_sector_size - (addr % current_sector_size), size);
-        // Need to use the internal page buffer in any of these two cases:
-        // 1. Size is not page aligned
-        // 2. Source buffer is not aligned to uint32_t. This is not supported by many targets (although
-        //    the pointer they accept is of uint8_t).
-        if (unaligned_src || (chunk < page_size)) {
-            chunk = std::min(chunk, page_size);
-            memcpy(_page_buf, buf, chunk);
-            if (chunk < page_size) {
-                memset(_page_buf + chunk, 0xFF, page_size - chunk);
+    // Few boards may fail the write actions due to HW limitations (like critical drivers that
+    // disable flash operations). Just retry a few times until success.
+    ret = -1;
+    for (unsigned int retry = 0; retry < num_write_retries && ret; retry++) {
```

Comment:
should be an exit trigger of retries loop in case of success (maybe if total_size == 0)?

Reply:
ret is the exit trigger.

Reply:
sorry, I missed ret at the condition.
Could flash_program_page() be called several times for one "program" ?
If so, what happens if  not the first one fails , will it retry from the beginning?

Reply:
Very good point. It will retry from the beginning. This will likely succeed in most of the flash components (programming the same value twice), if not all, but there's always a chance that this logic won't always work. Will try to modify the code accordingly.

Reply:
Modified the code to retry the physical operation (and not on the entire process). Please re-review.

Reply:
Liked the idea (-: